Funcky/Internal/ListWithSelector.cs

```diff
@@ -0,0 +1,57 @@
+using System.Collections;
+
+namespace Funcky.Internal
+{
+    internal class ListWithSelector<TSource, TResult> : IList<TResult>
+    {
+        private readonly IList<TSource> _source;
+        private readonly Func<TSource, int, TResult> _selector;
+
+        public ListWithSelector(IList<TSource> source, Func<IList<TSource>, Func<TSource, int, TResult>> selector)
+            => (_source, _selector) = (source, selector(source));
+
+        public int Count
+            => _source.Count;
+
+        public bool IsReadOnly
+            => true;
+
+        public TResult this[int index]
+        {
+            get => _selector(_source[index], index);
+            set => throw new InvalidOperationException();
```

Comment:
I think [`NotSupportedException`](https://docs.microsoft.com/en-us/dotnet/api/system.notsupportedexception?view=net-5.0) would be the correct exception when purposefully breaking contract.
(See the [`IList` implementation on `ImmutableList<T>`](https://github.com/dotnet/runtime/blob/87b928ab37d2b1fa2b7aeec93a76190e23a0bf61/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs#L898-L926))

Reply:
Fixed in 55206ff34bd90b262b6c37e35a1cc64498b1935d